Modules/_threadmodule.c

```diff
@@ -42,24 +42,58 @@ get_thread_state(PyObject *module)
 
 // _ThreadHandle type
 
+typedef enum {
+    THREAD_HANDLE_JOINED,
+    THREAD_HANDLE_DETACHED,
+} ThreadHandleState;
+
+// A handle to join or detach an OS thread.
+//
+// Joining or detaching the handle is idempotent; the underlying OS thread is
+// joined or detached only once. Concurrent operations block until it is their
+// turn to execute or an operation completes successfully. Once an operation
+// has completed successfully all future operations complete immediately.
 typedef struct {
     PyObject_HEAD
     struct llist_node node;  // linked list node (see _pythread_runtime_state)
+
+    // The `ident` and `handle` fields are immutable once the object is visible
+    // to threads other than its creator, thus they do not need to be accessed
+    // atomically.
     PyThread_ident_t ident;
     PyThread_handle_t handle;
-    char joinable;
+
+    // Set before a handle is accessible to any threads other than its creator
+    // and cleared post-fork. Does not need to be accessed atomically.
+    bool is_valid;
+
+    // Set immediately before `thread_run` returns to indicate that the OS
+    // thread is about to exit.
+    _PyEventRc *thread_is_exiting;
+
+    // State is set once by the first successful `join` or `detach` operation.
+    ThreadHandleState state;
+    _PyOnceFlag once;
```

Comment:
There are multiple members indicating progress/validity in this object: `state`, `is_valid`, `thread_is_exiting`.

IMHO it would be easier to reason about this with a single state member covering all (5?) possible states: invalid, running, exiting, joined, detached. It would also make sure we get transitions right, by adding runtime assertions.

(That said, the current logic looks ok)

Reply:
We need `thread_is_exiting` but let me see if I can collapse `state` and `is_valid`. 

Another thing that would simplify this quite a bit is if we could remove the `detach` method. AFAICT it's only used in tests (to test the logic around attempting to join a detached handle). Since detaching only happens in the destructor, I think it should be safe to remove. What do you think?

Reply:
Generally, we would rather not wait for the destructor to release resources, which is why `detach` is exposed (though it might not be called at higher levels currently?).

Reply:
> though it might not be called at higher levels currently?

Yes, it doesn't appear to be called anywhere outside of tests. If nothing is using it I think we should delete it.

Reply:
If `Thread` has a `close` method for deterministic resource management (and/or supported the context manager protocol), it could call the `detach` method. But we can perhaps remove it for now, indeed.